pkg/api/norman/customization/cluster/validator.go

```diff
@@ -534,3 +538,170 @@ func validateEKS(prevCluster, newCluster map[string]interface{}) error {
+	}
+
+	return nil
+}
+
+// validateGKECredentialAuth validates that a user has access to the credential they are setting and the credential
+// they are overwriting. If there is no previous credential such as during a create or the old credential cannot
+// be found, the auth check will succeed as long as the user can access the new credential.
+func validateGKECredentialAuth(request *types.APIContext, credential string, prevCluster *v3.Cluster) error {
+	var accessCred mgmtclient.CloudCredential
+	credentialErr := "error accessing cloud credential"
+	if err := access.ByID(request, &mgmtSchema.Version, mgmtclient.CloudCredentialType, credential, &accessCred); err != nil {
+		return httperror.NewAPIError(httperror.NotFound, credentialErr)
+	}
+
+	if prevCluster == nil {
+		return nil
+	}
+
+	if prevCluster.Spec.GKEConfig == nil {
+		return nil
+	}
+
+	// validate the user has access to the old cloud credential before allowing them to change it
+	credential = prevCluster.Spec.GKEConfig.GoogleCredentialSecret
+	if err := access.ByID(request, &mgmtSchema.Version, mgmtclient.CloudCredentialType, credential, &accessCred); err != nil {
+		if apiError, ok := err.(*httperror.APIError); ok {
+			if apiError.Code.Status == httperror.NotFound.Status {
+				// old cloud credential doesn't exist anymore, anyone can change it
+				return nil
+			}
+		}
+		return httperror.NewAPIError(httperror.NotFound, credentialErr)
+	}
+
+	return nil
+}
+
+// validateGKEKubernetesVersion checks whether a kubernetes version is provided and if it is supported
+func validateGKEKubernetesVersion(spec *v32.ClusterSpec, prevCluster *v3.Cluster) error {
+	clusterVersion := spec.GKEConfig.KubernetesVersion
+	if clusterVersion == nil {
+		return nil
+	}
+
+	if *clusterVersion == "" {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "cluster kubernetes version cannot be empty string")
+	}
+
+	return nil
+}
+
+// validateGKENodePools checks whether a given node pool version is empty or not supported.
+func validateGKENodePools(spec *v32.ClusterSpec) error {
+	nodepools := spec.GKEConfig.NodePools
+	if nodepools == nil {
+		return nil
+	}
+	if len(nodepools) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("must have at least one node pool"))
+	}
+
+	var errors []string
+
+	for _, np := range nodepools {
+		name := *np.Name
+		if name == "" {
+			return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("node pool name cannot be an empty string"))
+		}
+
+		version := np.Version
+		if version == nil || *version == "" {
+			errors = append(errors, fmt.Sprintf("node pool [%s] version cannot be empty", name))
+			continue
+		}
+	}
+
+	if len(errors) != 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf(strings.Join(errors, ";")))
+	}
+	return nil
+}
+
+func validateGKEClusterName(client v3.ClusterInterface, spec *v32.ClusterSpec) error {
```

Comment:
There is already a name-collision validation check for all clusters. A user cannot have two clusters with the exact same name: https://github.com/rancher/rancher/blob/9a3562666ec2022358153d2c7a9193964ec57db3/pkg/api/norman/store/cluster/cluster_store.go#L613